### PR TITLE
fix(image-stream): Fix Apple disk image detection & reading

### PR DESCRIPTION
**Pull request by jhermsmeier**
This fixes two things: The format detection, and a bug in `udif`.

First, by categorizing the `.dmg` extension as compressed image,
`.isSupportedImage()` would attempt to detect the format after stripping
the extension, causing it to be misdetected.

Second, `udif`'s ReadStream didn't add the `dataForkOffset` to its
position when reading blocks, causing the wrong data to be read for some images,
in turn causing zlib to error on invalid headers.

Changes:
- Classify `.dmg` as `type: 'image'`
- Update `udif` to 0.8.0

Change-Type: patch
Changelog-Entry: Fix Apple disk image detection & streaming

### CODE REVIEW

**Comment by jhermsmeier** [comment]
Tested it with a MacOS Sierra image, which features a non-zero `dataForkOffset`:

screen shot 2017-04-11 at 16 12 37

**Comment by lurch** [comment]
> First, by categorizing the .dmg extension as compressed image,
.isSupportedImage() would attempt to detect the format after stripping
the extension, causing it to be misdetected.

Hmmm, should that have been picked up by a unit-test? If not, could you please add a unit-test to prevent it again in future? :)

Maybe you could also add some notes into `lib/image-stream/supported.js` about the impact of the different `type` attributes?

**Comment by jhermsmeier** [comment]
> Maybe you could also add some notes into lib/image-stream/supported.js about the impact of the different type attributes?

Good idea! I've added notes to both the `README` and `supported.js`

File: lib/image-stream/README.md [hunk 1]
```diff
@@ -58,4 +58,11 @@ These are the rules for handling archive images:
 
 The module throws an error if the above rules are not met.
 
+Supported Formats
+-----------------
+
+There are currently three image types in supported formats: `image`, `compressed` and `archive`.
+Note that when marking an extension as `compressed`, the filename will be stripped of that extension,
+and the leftover extension examined to determine the compressed image format (i.e. `.img.gz -> .img`).
```
Comment (lurch) [inline lib/image-stream/README.md]: I guess this is a potential pitfall of identifying mime-types purely by file extension ;-)

Comment (lurch) [inline lib/image-stream/README.md]: Would be nice to also explain the differences between `image` and `archive`.

**Comment by lurch** [comment]
I still think an extra unit-test would be nice too.

File: lib/image-stream/supported.js [hunk 2]
```diff
@@ -16,6 +16,13 @@
 
 'use strict';
 
+/**
+ * Supported filename extensions
```
Comment (jviotti) [inline lib/image-stream/supported.js]: Can we use `@summary` for the title, and `@description` for the long comment? For example:

```js
/**
 * @summary Supported filename extensions
 * @description
 * NOTE: Extensions with type: 'compressed' will be stripped
 * ...
 */
```

File: lib/image-stream/README.md (hunk 1 above)
Comment (jviotti) [inline lib/image-stream/README.md]: Haha, yeah, that'd be a compressed archive containing an image.

File: lib/image-stream/supported.js (hunk 2 above)
Comment (jhermsmeier) [inline lib/image-stream/supported.js]: Yup, mended!

File: lib/image-stream/README.md (hunk 1 above)
Comment (jhermsmeier) [inline lib/image-stream/README.md]: Updated with added notes on all three types

**Comment by jhermsmeier** [comment]
> I still think an extra unit-test would be nice too.

I'm not sure how we could test against this – or what exactly we should test here. Any ideas?

**Comment by lurch** [comment]
> I'm not sure how we could test against this – or what exactly we should test here. Any ideas?

Okay, let me rephrase the question then...

_Why_ wasn't
> First, by categorizing the .dmg extension as compressed image,
.isSupportedImage() would attempt to detect the format after stripping
the extension, causing it to be misdetected.

picked up by any unit-tests?

**Comment by jhermsmeier** [comment]
Ah yes, because there are two places we check for image support – one in image-stream, and another in the GUI (Etcher.Models.SupportedFormats).

Added tests for all supported formats in `gui/models/supported-formats.spec` now, but that still won't save us from making that logical error in this setup.

File: lib/image-stream/README.md [hunk 3]
```diff
@@ -58,4 +58,19 @@ These are the rules for handling archive images:
...
+
+An extension tagged `image` describes a format which can be directly written to a device by its handler,
+and an extension tagged `archive` denotes an archive containing an image, and will cause an archive handler
+to open the archive and search for an image file.
+
+As an archive (such as `.tar`) might be additionally compressed, this will allow for constructs such as
+`.tar.gz` (a compressed archive, containing a file with an extension tagged as `image`) to be handled correctly.
+
+Note that when marking an extension as `compressed`, the filename will be stripped of that extension,
+and the leftover extension examined to determine the compressed image format (i.e. `.img.gz -> .img`).
```
Comment (lurch) [inline lib/image-stream/README.md]: Might read better with this sentence moved before the previous one?

File: lib/image-stream/supported.js [hunk 4]
```diff
@@ -16,6 +16,14 @@
 
 'use strict';
 
+/**
+ * @summary Supported filename extensions
+ * @description
+ * NOTE: Extensions with type: 'compressed' will be stripped
+ * from filenames to determine the format of the compressed image.
```
Comment (lurch) [inline lib/image-stream/supported.js]: Perhaps change "of the compressed image." to "of the uncompressed image." ?

File: tests/gui/models/supported-formats.spec.js [hunk 5]
```diff
@@ -61,6 +62,17 @@ describe('Browser: SupportedFormats', function() {
 
     describe('.isSupportedImage()', function() {
 
+      _.forEach(imageStream.supportedFileTypes, (fileType) => {
```
Comment (lurch) [inline tests/gui/models/supported-formats.spec.js]: To match up with the style of the other unit tests, I think this should be explicitly spelled out, rather than just imported from `imageStream` ?  @jviotti to confirm :)

(and would that also prevent the "but that still won't save us from making that logical error in this setup" you refer to?)

File: lib/image-stream/README.md (hunk 3 above)
Comment (jhermsmeier) [inline lib/image-stream/README.md]: Done

File: lib/image-stream/supported.js (hunk 4 above)
Comment (jhermsmeier) [inline lib/image-stream/supported.js]: Fixed

File: tests/gui/models/supported-formats.spec.js (hunk 5 above)
Comment (jhermsmeier) [inline tests/gui/models/supported-formats.spec.js]: Done. I guess it'll at least be another barrier at which one could notice.

File: lib/image-stream/README.md [hunk 6]
```diff
@@ -58,4 +58,19 @@ These are the rules for handling archive images:
...
 The module throws an error if the above rules are not met.
 
+Supported Formats
+-----------------
+
+There are currently three image types in supported formats: `image`, `compressed` and `archive`.
+
+An extension tagged `image` describes a format which can be directly written to a device by its handler,
+and an extension tagged `archive` denotes an archive containing an image, and will cause an archive handler
+to open the archive and search for an image file.
```
Comment (lurch) [inline lib/image-stream/README.md]: Ridiculous-hypothetical-question: If a compressed image was inside a zip, which was inside another zip, would that transparently "just work"? ;-)

File: lib/image-stream/README.md [hunk 7]
```diff
@@ -58,4 +58,19 @@ These are the rules for handling archive images:
...
+
+Note that when marking an extension as `compressed`, the filename will be stripped of that extension,
+and the leftover extension examined to determine the compressed image format (i.e. `.img.gz -> .img`).
```
Comment (lurch) [inline lib/image-stream/README.md]: As with my earlier comment, I guess this could be "to determine the uncompressed image format" ?

File: tests/gui/models/supported-formats.spec.js [hunk 8]
```diff
@@ -61,6 +61,32 @@ describe('Browser: SupportedFormats', function() {
...
+        'path/to/filename.img.bz2',
+        'path/to/filename.img.xz',
+
+        // Type: 'image'
+        'path/to/filename.img',
+        'path/to/filename.iso',
+        'path/to/filename.dsk',
+        'path/to/filename.hddimg',
+        'path/to/filename.raw',
+        'path/to/filename.dmg'
```
Comment (lurch) [inline tests/gui/models/supported-formats.spec.js]: Would it be worth it / possible to check that all these filenames resolve to the correct expected mime-type?
(or is that already covered elsewhere?)

P.S. Looking at https://github.com/resin-io/etcher/pull/1283/files#diff-dd8adcb7744d5236c18272b3e227071eR144 I can now see why the problem that this PR fixes wasn't picked up by the unit-tests - they only check the _first_ of each supported extension type, so of course `.dmg` didn't get tested because it was on the end of the list.

File: tests/gui/models/supported-formats.spec.js [hunk 9]
```diff
@@ -61,6 +61,32 @@ describe('Browser: SupportedFormats', function() {
...
+
+      ], (filename) => {
+        it(`should return true for ${filename}`, function() {
+          const isSupported = SupportedFormatsModel.isSupportedImage(filename);
```
Comment (lurch) [inline tests/gui/models/supported-formats.spec.js]: :+1:

File: tests/gui/models/supported-formats.spec.js (hunk 8 above)
Comment (jhermsmeier) [inline tests/gui/models/supported-formats.spec.js]: That's already covered in the image-stream specs

File: lib/image-stream/README.md (hunk 6 above)
Comment (jhermsmeier) [inline lib/image-stream/README.md]: Nope, because Etcher only checks one level deep in archives, and yells if it can't find a suitable image

File: lib/image-stream/README.md (hunk 7 above)
Comment (jhermsmeier) [inline lib/image-stream/README.md]: Mended

**Comment by jviotti** [comment]
I approved, but should we add a `Changelog-Entry` footer tag describing the fix?

**Comment by jhermsmeier** [comment]
@jviotti Added
